legacy/expressions/visitor.go

```diff
@@ -193,50 +178,35 @@ func (v *legacyVisitor) VisitAdditionOrSubtraction(ctx *gen.AdditionOrSubtractio
 		op = "-"
 	}
 
-	secondIsDate := isDate(next)
-	if secondIsDate {
-		secondSeconds, ok := convertTimeToSeconds(next)
-		if ok {
-			next = secondSeconds
-			dateUnit = "s"
-		}
-	}
+	// see if either of our arguments is a date value
+	arg1Type := inferType(arg1)
+	arg2Type := inferType(arg2)
 
-	// see if our second param is an int
-	_, secondNumberErr := strconv.Atoi(next)
-	if (firstIsDate || secondIsDate) && (firstNumberErr != nil || secondNumberErr != nil) {
+	//fmt.Printf("Migrating add/sub with types: %s => %s, %s =>%s\n", arg1, arg1Type, arg2, arg2Type)
 
-		// we are adding two values where we know at least one side is a date
-		template := "datetime_add(%s, %s, \"%s\")"
-		if op == "-" {
-			template = "datetime_add(%s, -%s, \"%s\")"
-		}
+	if arg1Type == "number" && arg2Type == "number" {
+		// we are adding two numbers
+		return fmt.Sprintf("%s %s %s", arg1, op, arg2)
 
-		// determine the order of our parameters
-		replacements := []interface{}{value, next, dateUnit}
-		if firstNumberErr == nil {
-			replacements = []interface{}{next, value, dateUnit}
+	} else if arg1Type == "datetime" && arg2Type == "number" {
+		// we are adding a date and a number (of days)
+		template := `datetime_add(%s, %s, "D")`
+		if op == "-" {
+			template = `datetime_add(%s, -%s, "D")`
 		}
 
-		value = fmt.Sprintf(template, replacements...)
+		return fmt.Sprintf(template, arg1, arg2)
 
-	} else if firstNumberErr == nil && secondNumberErr == nil {
-		// we are adding two numbers
-		if op == "+" {
-			value = fmt.Sprintf("%s + %s", value, next)
-		} else {
-			value = fmt.Sprintf("%s - %s", value, next)
-		}
-	} else {
-		// we are adding a field of unknown type with an integer
-		if op == "+" {
-			value = fmt.Sprintf("legacy_add(%s, %s)", value, next)
-		} else {
-			value = fmt.Sprintf("legacy_add(%s, -%s)", value, next)
-		}
+	} else if arg1Type == "datetime" && arg2Type == "time" && op == "+" {
+		// we are adding a date and a time
+		return fmt.Sprintf(`replace_time(%s, %s)`, arg1, arg2)
```

Comment:
making a bit of an assumption here that anyone trying to add a date and a time actually wants to replace rather than add the time value...